### PR TITLE
Extract Navigation Block data fetching and management to hook(s)

### PR DESCRIPTION
**Pull request by getdave**
## Description
Currently the logic for fetching and management of data (eg: Pages, Menus, Menu Items) in the Nav Block is placed inside the `placeholder.js` component. As we look to extract reusable logic ([possibly into a `@wordpress/menus` package](https://github.com/WordPress/gutenberg/issues/31580)) it's important that important logic such as this is not "buried" within specific components.

Moreover, due to being placed inline within a component the data logic is currently all combined and difficult to read / comprehend.

To this end this PR breaks the data fetching out into separate (potentially reusable) hooks for each entity whilst providing a master utility hook `useNavigationEntities` which will pull through all entity data in one go.

The benefits of this are:

- Makes the entity fetching logic easier to comprehend and understand.
- Better separation of concerns.
- Easier to extract and refactor to `@wordpress/menus` if required in future (aligns more closely with [the Nav Editor data fetching implementation](https://github.com/WordPress/gutenberg/blob/trunk/packages/edit-navigation/src/hooks/use-navigation-editor.js))

## How has this been tested?

Check all existing functionality is "as is". Nothing should change.

Particularly check that

* Nav Block placeholder state and functionality continues to behave correctly in Post/Site Editor.
* Nav Editor continues to work.

## Types of changes
New feature (non-breaking change which adds functionality) 

## Checklist:
- [ ] My code is tested.
- [ ] My code follows the WordPress code style. 
- [ ] My code follows the accessibility standards. 
- [ ] I've tested my changes with keyboard and screen readers. 
- [ ] My code has proper inline documentation. 
- [ ] I've included developer documentation if appropriate. 
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal).

### CODE REVIEW

File: packages/block-library/src/navigation/use-navigation-entities.js [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+					: undefined,
+				hasResolvedMenuItems: hasSelectedMenu
+					? hasFinishedResolution(
+							'getMenuItems',
+							menuItemsParameters
+					  )
+					: false,
+			};
+		},
+		[ menuId ]
```
Comment (getdave) [inline packages/block-library/src/navigation/use-navigation-entities.js:92]: Question: will this selector be retriggered when the selected menu ID (passed from `placeholder.js`) changes?

Comment (Mamaduka) [reply packages/block-library/src/navigation/use-navigation-entities.js:92]: I think so. This should be similar to `useEffect` dependencies.

File: packages/block-library/src/navigation/use-navigation-entities.js [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+
+		return {
+			menus: getMenus( ...menusParameters ),
+			isResolvingMenus: isResolving( 'getMenus', menusParameters ),
+			hasResolvedMenus: hasFinishedResolution(
+				'getMenus',
+				menusParameters
+			),
+		};
+	} );
```
Comment (Mamaduka) [inline packages/block-library/src/navigation/use-navigation-entities.js]: Let's add an empty dependency array to optimize performance. Similar work in the past #23255.

File: packages/block-library/src/navigation/use-navigation-entities.js [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+					'getEntityRecords',
+					pagesParameters
+				),
+				hasResolvedPages: hasFinishedResolution(
+					'getEntityRecords',
+					pagesParameters
+				),
+			};
+		}
+	);
```
Comment (Mamaduka) [inline packages/block-library/src/navigation/use-navigation-entities.js:134]: We can do similar optimization here.

File: packages/block-library/src/navigation/use-navigation-entities.js [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+		};
+	} );
+
+	return {
+		menus,
+		isResolvingMenus,
+		hasResolvedMenus,
+		menuItems,
+		hasResolvedMenuItems,
+		hasMenus: !! ( hasResolvedMenus && menus?.length ),
```
Comment (Mamaduka) [inline packages/block-library/src/navigation/use-navigation-entities.js:61]: I think we can move this calculation inside the `useSelect` callback.

File: packages/block-library/src/navigation/use-navigation-entities.js [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+				),
+			};
+		}
+	);
+
+	return {
+		pages,
+		isResolvingPages,
+		hasResolvedPages,
+		hasPages: !! ( hasResolvedPages && pages?.length ),
```
Comment (Mamaduka) [inline packages/block-library/src/navigation/use-navigation-entities.js:140]: Same here.

File: packages/block-library/src/navigation/use-navigation-entities.js [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+	};
+}
+
+function useMenuEntities() {
+	const {
+		menus,
+		isResolvingMenus,
+		hasResolvedMenus,
+		menuItems,
+		hasResolvedMenuItems,
```
Comment (Mamaduka) [inline packages/block-library/src/navigation/use-navigation-entities.js]: `menuItems` and `hasResolvedMenuItems` aren't returned from the `useSelect` callback, so they always be `undefined.`

Comment (getdave) [reply packages/block-library/src/navigation/use-navigation-entities.js]: Whoops. Left over from refactor.

File: packages/block-library/src/navigation/use-navigation-entities.js (hunk 4 above)
Comment (getdave) [reply packages/block-library/src/navigation/use-navigation-entities.js:61]: Looks like doing this causes an error with mapSelect. Is there a major downside to doing it the way I have it now I've restored to the original?

File: packages/block-library/src/navigation/use-navigation-entities.js (hunk 5 above)
Comment (getdave) [reply packages/block-library/src/navigation/use-navigation-entities.js:140]: See https://github.com/WordPress/gutenberg/pull/31825#discussion_r632470572

File: packages/block-library/src/navigation/use-navigation-entities.js (hunk 4 above)
Comment (Mamaduka) [reply packages/block-library/src/navigation/use-navigation-entities.js:61]: No major downside in this case. It's just a practice I picked up from the core.

**Review by Mamaduka — APPROVED** [review]
LGTM :shipit:

I couldn't spot any regressions for Navigation Block in Post/Site Editor, and Navigation Screen works as expected.

**Pull request merged**